plugins/error-augmentor/plugin.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright © 2021 Aspect Build Systems Inc
+
+Not licensed for re-use.
+*/
+
+package main
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	goplugin "github.com/hashicorp/go-plugin"
+	"gopkg.in/yaml.v2"
+
+	buildeventstream "aspect.build/cli/bazel/buildeventstream/proto"
+	"aspect.build/cli/pkg/ioutils"
+	"aspect.build/cli/pkg/plugin/sdk/v1alpha2/config"
+)
+
+func main() {
+	goplugin.Serve(config.NewConfigFor(NewDefaultPlugin()))
+}
+
+type ErrorAugmentorPlugin struct {
+	properties          string
+	regex               map[*regexp.Regexp]string
+	helpfulHints        []string
+	yamlUnmarshalStrict func(in []byte, out interface{}) (err error)
+}
+
+func NewDefaultPlugin() *ErrorAugmentorPlugin {
+	return NewPlugin()
+}
+
+func NewPlugin() *ErrorAugmentorPlugin {
+	return &ErrorAugmentorPlugin{
+		properties:          "",
+		regex:               map[*regexp.Regexp]string{},
+		helpfulHints:        make([]string, 0),
+		yamlUnmarshalStrict: yaml.UnmarshalStrict,
+	}
+}
+
+// will be used to unmarshal plugin properties specific to this plugin
+type errorMappings struct {
+	ErrorMappings map[string]string `yaml:"error_mappings"`
+}
+
+func (plugin *ErrorAugmentorPlugin) SetupHook(
+	properties string,
+) error {
+	plugin.properties = properties
+
+	var processedProperties errorMappings
+	if err := plugin.yamlUnmarshalStrict([]byte(properties), &processedProperties); err != nil {
+		return fmt.Errorf("failed to parse incoming properties: %w", err)
+	}
+
+	// change map keys into regex objects now so they are ready to use and we only need to compile the regex once
+	for r, m := range processedProperties.ErrorMappings {
+		// r for regex, m for message
+		plugin.regex[regexp.MustCompile(r)] = m
+	}
+
+	return nil
+}
+
+func (plugin *ErrorAugmentorPlugin) BEPEventCallback(event *buildeventstream.BuildEvent) error {
+	aborted := event.GetAborted()
+	if aborted != nil {
+		plugin.processErrorMessage(aborted.Description)
+
+		// if aborted then there will not be a progress messages so exit early
+		return nil
+	}
+
+	progress := event.GetProgress()
+
+	if progress != nil {
+		stderr := progress.GetStderr()
+		plugin.processErrorMessage(stderr)
```

Comment:
I think you want at least to check if `stderr` is not empty before calling `processErrorMessage`. That method is quite expensive to run on every event coming from Bazel.

Ideally, we should be able to do a check on the event to see if it represents an error. Could you check this? If not, take my following suggestion:

```suggestion
		if stderr != "" {
			plugin.processErrorMessage(stderr)
		}
```

Reply:
You could also spawn a goroutine as the error message processor and communicate via non-blocking channels, so Bazel doesn't wait for you to do all the regex matching before it proceeds. I can help pairing with you on this.